### PR TITLE
torch: handle `torch.prim.dtype` ops during type refinement

### PR DESCRIPTION
**Pull request by ashay**
The canonicalizer converts `torch.prim.dtype` ops into integer constants
for valid types, but the type may not be known until type refinement is
complete.  However, type refinement cannot make progress until
`torch.prim.dtype` ops have been resolved to their corresponding integer
constants, thus creating a circular dependency.

This patch creates a tight coupling between type refinement and the
lowering of `torch.prim.dtype` ops by handling such ops as they are
encountered during type refinement.  The unit test in this patch aims to
check whether the type refinement pass can now handle chains of
operations that alternate between type construction and type refinement.

### CODE REVIEW

File: lib/Dialect/Torch/Transforms/RefineTypes.cpp [hunk 1]
```diff
@@ -587,24 +598,39 @@ getPromotedResultTypeAssumingNonZeroRank(MLIRContext *context,
...
+  else if (auto primDtypeOp = dyn_cast<PrimDtypeOp>(dtype.getDefiningOp())) {
+    auto element = getLatticeElement(primDtypeOp.a());
+    assert(!element.isUninitialized() && "no refined type for source");
+
+    switch (element.getValue().kind) {
+    case torch_upstream::TypeKind::TensorType:
+      knowledge.dtype = element.getValue().dtype;
+      break;
+
+    case torch_upstream::TypeKind::NumberType:
```
Comment (silvasean) [inline lib/Dialect/Torch/Transforms/RefineTypes.cpp:618]: do we need to handle this case? I don't think it is legal to call prim.dtype on a scalar value.

File: lib/Dialect/Torch/Transforms/RefineTypes.cpp [hunk 2]
```diff
@@ -587,24 +598,39 @@ getPromotedResultTypeAssumingNonZeroRank(MLIRContext *context,
...
+    ValueKnowledge &knowledge, Value dtype, Type inputDType) {
   assert(isBuiltInType(inputDType) && "`inputDType` must be a builtin type");
   int64_t dtypeInt;
   if (dtype.getType().isa<Torch::NoneType>())
     knowledge.dtype = inputDType;
   else if (matchPattern(dtype, m_TorchConstantInt(&dtypeInt)))
     knowledge.dtype = getTypeForDTypeInteger(dtype.getContext(), dtypeInt);
+  else if (auto primDtypeOp = dyn_cast<PrimDtypeOp>(dtype.getDefiningOp())) {
+    auto element = getLatticeElement(primDtypeOp.a());
+    assert(!element.isUninitialized() && "no refined type for source");
```
Comment (silvasean) [inline lib/Dialect/Torch/Transforms/RefineTypes.cpp:611]: is it possible to write a test case to hit this, or is it actually "impossible"?

File: lib/Dialect/Torch/Transforms/RefineTypes.cpp (hunk 1 above)
Comment (ashay) [reply lib/Dialect/Torch/Transforms/RefineTypes.cpp:618]: You're right, `prim.dtype` should never have a scalar operand.  I'll remove the case.

File: lib/Dialect/Torch/Transforms/RefineTypes.cpp (hunk 2 above)
Comment (ashay) [reply lib/Dialect/Torch/Transforms/RefineTypes.cpp:611]: That's a good point, let me think about it.

Comment (ashay) [reply lib/Dialect/Torch/Transforms/RefineTypes.cpp]: Thanks, that makes sense.  Removing that assertion and the unnecessary cases greatly simplified the code!

File: lib/Dialect/Torch/Transforms/RefineTypes.cpp (hunk 1 above)
Comment (ashay) [reply lib/Dialect/Torch/Transforms/RefineTypes.cpp]: Removed.

**Review by silvasean — APPROVED** [review]
Great!

**Pull request merged**